### PR TITLE
[3.11] Improve locking in Async Logger (#19989)

### PR DESCRIPTION
**Pull request by mchacki**
### Scope & Purpose

*Right now the Logger holds a very large lock. The lock is held whenever the background thread flushes a queued log message, which is potentially slowed down by IO. If we try to do a new direct log entry in that period we need to wait for this lock to be released. With this PR the Lock is changed such that: We get a WriteLock whenever we modify the Loggers we use, but we only take a read-lock when we let the loggers print. This way we can have two or more threads using the loggers.
In turn the output message of the Logger, which does the IO operation, is guarded with a WriteLock.

In addition this PR adds an upper bound to the amount of messages in the queue. This is right now hard-coded and should not have effect under normal circumstances. If the logging is configured to use an extremely slow device we protect us here against accumulating too much memory in log messages. If the Limit is reached, we will start to directly log the message, which will have a performance impact. Also this may change the ordering of Log messages.
*

Note to reviewer: Please take extra care, this PR fiddles with Locks, and has potential to introduce undesired dead-locks.

This potentially fixes: https://arangodb.atlassian.net/browse/BTS-1273

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/BTS-1273
- [ ] Design document:

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -5,6 +5,24 @@ v3.11.5 (XXXX-XX-XX)
...
+  to grow indefinitely. The upper bound is configurable via the startup
+  option `--log.max-queued-entries`. The default value is 10000.
+
+* Silence TSAN for shutdown for access to the SchedulerFeature::SCHEDULER
+  pointer using atomic references.
+
+* Fixed a race in controlled leader change which could lead to a situation 
+  in which a shard follower is dropped when the first write operation
+  happens. This fixes BTS-1647.
+
```
Comment (jsteemann) [inline CHANGELOG:25]: These comments do not seem to belong to this PR.

Comment (jsteemann) [inline CHANGELOG:25]: These comments do not seem to belong to this PR.
```suggestion
```

File: lib/Logger/LoggerFeature.cpp [hunk 2]
```diff
@@ -556,6 +556,21 @@ printed in the thread that triggered the log message. This is non-optimal for
...
 off to an extra logging thread, which asynchronously writes the log messages.)");
 
+  options
+      ->addOption(
+          "--log.max-queued-entries",
+          "Upper limit of log entries that are queued in background thread.",
+          new UInt32Parameter(&_maxQueuedLogMessages),
+          arangodb::options::makeDefaultFlags(
+              arangodb::options::Flags::Uncommon))
+      .setIntroducedIn(31200)
```
Comment (jsteemann) [inline lib/Logger/LoggerFeature.cpp:566]: May be adjusted for 3.11.
```suggestion
      .setIntroducedIn(31105)
```

**Review by jsteemann — APPROVED** [review]
LGTM

**Comment by mchacki** [comment]
Jenkins green:
https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/32037/

**Review by neunhoef — APPROVED** [review]
LGTM
